lib/class-wp-block.php

```diff
@@ -0,0 +1,234 @@
+					$this->context[ $context_name ] = $this->available_context[ $context_name ];
+				}
+			}
+		}
+
+		if ( ! empty( $block['innerBlocks'] ) ) {
+			$child_context = $this->available_context;
+
+			/* phpcs:disable WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase */
+			if ( ! empty( $this->block_type->providesContext ) ) {
+				foreach ( $this->block_type->providesContext as $context_name => $attribute_name ) {
+					if ( isset( $this->attributes[ $attribute_name ] ) ) {
+						$child_context[ $context_name ] = $this->attributes[ $attribute_name ];
+					}
+				}
+			}
+			/* phpcs:enable */
+
+			$this->inner_blocks = array_map(
+				function( $inner_block ) use ( $child_context, $registry ) {
+					return new WP_Block( $inner_block, $child_context, $registry );
+				},
+				$block['innerBlocks']
+			);
+		}
+
+		if ( ! empty( $block['innerHTML'] ) ) {
+			$this->inner_html = $block['innerHTML'];
+		}
+
+		if ( ! empty( $block['innerContent'] ) ) {
+			$this->inner_content = $block['innerContent'];
+		}
+	}
+
+	/**
+	 * Generates the render output for the block.
+	 *
+	 * @return string Rendered block output.
+	 */
+	public function render() {
+		global $post;
+
+		$is_dynamic    = $this->name && null !== $this->block_type && $this->block_type->is_dynamic();
+		$block_content = '';
+
+		$index = 0;
+		foreach ( $this->inner_content as $chunk ) {
+			$block_content .= is_string( $chunk ) ?
+				$chunk :
+				$this->inner_blocks[ $index++ ]->render();
+		}
+
+		if ( $is_dynamic ) {
+			$global_post   = $post;
+			$block_content = (string) call_user_func( $this->block_type->render_callback, $this, $block_content );
+			$post          = $global_post;
+		}
+
+		return $block_content;
+	}
+
+	/**
+	 * Returns true if an attribute exists by the specified attribute name, or
+	 * false otherwise.
+	 *
+	 * @link https://www.php.net/manual/en/arrayaccess.offsetexists.php
+	 *
+	 * @param string $attribute_name Name of attribute to check.
+	 *
+	 * @return bool Whether attribute exists.
+	 */
+	public function offsetExists( $attribute_name ) {
+		return isset( $this->attributes[ $attribute_name ] );
+	}
+
+	/**
+	 * Returns the value by the specified attribute name.
+	 *
+	 * @link https://www.php.net/manual/en/arrayaccess.offsetget.php
+	 *
+	 * @param string $attribute_name Name of attribute value to retrieve.
+	 *
+	 * @return mixed|null Attribute value if exists, or null.
+	 */
+	public function offsetGet( $attribute_name ) {
+		return isset( $this->attributes[ $attribute_name ] ) ?
+			$this->attributes[ $attribute_name ] :
+			null;
+	}
+
+	/**
+	 * Assign an attribute value by the specified attribute name.
+	 *
+	 * @link https://www.php.net/manual/en/arrayaccess.offsetset.php
+	 *
+	 * @param string $attribute_name Name of attribute value to set.
+	 * @param mixed  $value          Attribute value.
+	 */
+	public function offsetSet( $attribute_name, $value ) {
+		if ( ! is_null( $attribute_name ) ) {
```

Comment:
It looks like this guards against setting new values like `$attributes[] = 'some value';` would.

Is there a concern about the backwards compatibility implications of that? Granted that assigning to attributes like that is not really recommended.

I think there are two choices, either guarantee that the behavior is exactly what it was for an array, or add some more protection around array access. I think in the latter case the documentation should mention the difference, but I also think a warning/error might be considered.

Reply:
> It looks like this guards against setting new values like `$attributes[] = 'some value';` would.
> 
> Is there a concern about the backwards compatibility implications of that? Granted that assigning to attributes like that is not really recommended.
> 
> I think there are two choices, either guarantee that the behavior is exactly what it was for an array, or add some more protection around array access. I think in the latter case the documentation should mention the difference, but I also think a warning/error might be considered instead of the silent failure.

Also a good observation, and very similar to https://github.com/WordPress/gutenberg/pull/21467#discussion_r410004286 in terms of what you note about aligning to expected behaviors of an array.

The [documentation example](https://www.php.net/manual/en/class.arrayaccess.php) I linked in my previous comment is relevant here as well, since it _does_ include the "push" semantics. In implementing this, I had gone out of my way to disallow this. But upon your comment and some reflection, I do think it should behave identically to an array, especially in mind of backward-compatibility, and despite the fact that it may not be a behavior we strictly _want_ to support.

Reply:
> It looks like this guards against setting new values like `$attributes[] = 'some value';` would.
> 
> Is there a concern about the backwards compatibility implications of that? Granted that assigning to attributes like that is not really recommended.
> 
> I think there are two choices, either guarantee that the behavior is exactly what it was for an array, or add some more protection around array access. I think in the latter case the documentation should mention the difference, but I also think a warning/error might be considered instead of the silent failure.

Updated in f8beb9e (toward the first of your two presented options).